lib/libuzfs/libuzfs.c

```diff
@@ -0,0 +1,740 @@
+    nvlist_t *fsprops)
+{
+	nvlist_t *nvroot;
+	int err = 0;
+
+	nvroot = make_vdev_root(path, NULL, zpool, 0, 0, NULL, 1, 0, 1);
+
+	err = spa_create(zpool, nvroot, props, NULL, NULL);
+	if (err) {
+		goto out;
+	}
+
+out:
+	fnvlist_free(nvroot);
+	return (err);
+}
+
+int
+libuzfs_zpool_destroy(const char *zpool)
+{
+	return (spa_destroy(zpool));
+}
+
+libuzfs_zpool_handle_t *
+libuzfs_zpool_open(const char *zpool)
+{
+	int err = 0;
+	spa_t *spa;
+	err = spa_open(zpool, &spa, FTAG);
+	if (err)
+		return (NULL);
+
+	libuzfs_zpool_handle_t *zhp;
+	zhp = umem_alloc(sizeof (libuzfs_zpool_handle_t), UMEM_NOFAIL);
+	zhp->spa = spa;
+	(void) strlcpy(zhp->zpool_name, zpool, sizeof (zhp->zpool_name));
+
+	return (zhp);
+}
+
+void
+libuzfs_zpool_close(libuzfs_zpool_handle_t *zhp)
+{
+	spa_close(zhp->spa, FTAG);
+	free(zhp);
+}
+
+void
+libuzfs_zpool_prop_set(libuzfs_zpool_handle_t *zhp, zpool_prop_t prop,
+    uint64_t value)
+{
+	libuzfs_spa_prop_set_uint64(zhp->spa, prop, value);
+}
+
+int
+libuzfs_zpool_prop_get(libuzfs_zpool_handle_t *zhp, zpool_prop_t prop,
+    uint64_t *value)
+{
+	int err = 0;
+	nvlist_t *props = NULL;
+
+	VERIFY0(spa_prop_get(zhp->spa, &props));
+
+	nvlist_t *propval;
+	err = nvlist_lookup_nvlist(props, zpool_prop_to_name(prop), &propval);
+	if (err) {
+		goto out;
+	}
+
+	*value = fnvlist_lookup_uint64(propval, ZPROP_VALUE);
+
+out:
+	fnvlist_free(props);
+	return (err);
+}
+
+static void
+libuzfs_objset_create_cb(objset_t *os, void *arg, cred_t *cr, dmu_tx_t *tx)
+{
+	/*
+	 * Create the objects common to all libuzfs datasets.
+	 */
+	VERIFY0(zap_create_claim(os, LIBUZFS_DIROBJ, DMU_OT_ZAP_OTHER,
+	    DMU_OT_NONE, 0, tx));
+}
+
+int
+libuzfs_dataset_create(const char *dsname)
+{
+	int err;
+
+	err = dmu_objset_create(dsname, DMU_OST_ZFS, 0, NULL,
+	    libuzfs_objset_create_cb, NULL);
+	if (err)
+		return (err);
+
+	return (libuzfs_dsl_prop_set_uint64(dsname, ZFS_PROP_SYNC,
+	    ZFS_SYNC_ALWAYS, B_FALSE));
+}
+
+static int
+libuzfs_objset_destroy_cb(const char *name, void *arg)
+{
+	objset_t *os;
+	dmu_object_info_t doi;
+	int err;
+
+	/*
+	 * Verify that the dataset contains a directory object.
+	 */
+	VERIFY0(libuzfs_dmu_objset_own(name, DMU_OST_ZFS, B_TRUE, B_TRUE, FTAG,
+	    &os));
+	err = dmu_object_info(os, LIBUZFS_DIROBJ, &doi);
+	if (err != ENOENT) {
+		/* We could have crashed in the middle of destroying it */
+		ASSERT0(err);
+		ASSERT3U(doi.doi_type, ==, DMU_OT_ZAP_OTHER);
+		ASSERT3S(doi.doi_physical_blocks_512, >=, 0);
+	}
+	dmu_objset_disown(os, B_TRUE, FTAG);
+
+	/*
+	 * Destroy the dataset.
+	 */
+	if (strchr(name, '@') != NULL) {
+		VERIFY0(dsl_destroy_snapshot(name, B_TRUE));
+	} else {
+		err = dsl_destroy_head(name);
+		if (err != EBUSY) {
+			/* There could be a hold on this dataset */
+			ASSERT0(err);
+		}
+	}
+	return (0);
+}
+
+void
+libuzfs_dataset_destroy(const char *dsname)
+{
+	(void) dmu_objset_find(dsname, libuzfs_objset_destroy_cb, NULL,
```

Comment:
just a find?

Reply:
find will call the parameter: libuzfs_objset_destroy_cb to do real work.